### PR TITLE
Proposal: Changelog Automation Updates

### PR DESCRIPTION
**Pull request by filiptronicek**
### TL;DR

- the changelog has a different file structure (one folder for each)
- a new format for release notes, backwards-compatible with the old release note blocks
- there are changes to fully automate the changelog process (GitHub workflows for automatic updating of changelog PRs)
- the changelog respects deployed status
- the PRs in the changelog are now automatically categorized
- we now fetch from the `workspace-images` & the `website` repo as well as `gitpod`

## The new changelog

The changelog has been re-made to reflect many challenges we had with it. 

The first thing that has changed is the file structure of the changelogs - they are no longer single `.md` files, but rather multiple markdowns in a folder called after the release date of the changelog (usually the last business day of the month).

The command for generating changelogs also got simpler, now it is as simple as:
```sh
node scripts/changelog
```

> **Note**: a pre-requisite to running the command is to have the `CHANGELOG_GITHUB_ACCESS_TOKEN` environment variable set-up with a [GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token). The token can also be supplied via the `token` flag.

<details>
<summary>Sample command output</summary>

```
Hello, Filip Troníček

Fetching PRs for gitpod-io/gitpod from 2022-08-04 to 2022-10-31
Search query: repo:gitpod-io/gitpod is:pr is:merged merged:2022-08-04..2022-10-31 sort:updated-desc label:deployed -label:release-note-none -project:gitpod-io/22
Found 14 PRs after filtering for gitpod-io/gitpod (from 81 total)
[server]  Disallow removing all Owners from a team is categorized as Server but it doesn't have the label component: server
[dashboard] Fetch user BillingMode on load (or team change) is categorized as Dashboard but it doesn't have the label component: dashboard
[dashboard] Deprecate the creation of projects under individual accounts is categorized as Dashboard but it doesn't have the label component: dashboard
Fetching PRs for gitpod-io/website from 2022-10-03 to 2022-10-31
Search query: repo:gitpod-io/website is:pr is:merged merged:2022-10-03..2022-10-31 sort:updated-desc
Found 36 PRs after filtering for gitpod-io/website (from 36 total)
Fetching PRs for gitpod-io/workspace-images from 2022-10-03 to 2022-10-31
Search query: repo:gitpod-io/workspace-images is:pr is:merged merged:2022-10-03..2022-10-31 sort:updated-desc
Found 5 PRs after filtering for gitpod-io/workspace-images (from 5 total)
Changelog generated. Please edit ./src/lib/contents/changelog/2022-10-31/index.md
```

</details>

The changelog file itself now looks like this:

```md

{manually added content in the begging of the changelog}

<!--- BEGIN_AUTOGENERATED_CHANGES -->

## Category 1

{PARTIAL}

- PR 1
- PR 2
- PR 3

## Category 2

- PR 1

## Fixes and improvements

- PR 1
- PR 2
- PR 3
- PR 4


<!--- END_AUTOGENERATED_CHANGES -->

```
### QnA

**What's a `Category`?**

A category is a group of PRs that belong together in the changelog (e.g. VS Code / Dashboard).  [`config.js`](https://github.com/gitpod-io/website/blob/3e0ccf380d1c1d8d790589d1fe2073deaf4c030a/scripts/changelog/lib/config.js) is the place housing the definitions of what type of PRs should go into any given category (what label, title prefix it has to have)

**What's a `Partial`?**

A partial is a separate file for adding ad-hoc content to the changelog, belonging to a specific `Category`.

By default, there are no partials in a given changelog. To create one, first ensure that you have the changelog folder and index file already generated, and then, in the same folder create a file named after your partial's desired category's `partial` property (can again be found in [`config.js`](https://github.com/gitpod-io/website/blob/3e0ccf380d1c1d8d790589d1fe2073deaf4c030a/scripts/changelog/lib/config.js)). 

This means that for instance, if you wanted to add some more content about a big VS Code feature to this month's changelog, you would go to `2022-10-31/` and create the `vscode.md` file there. Then, you would populate it with any content you desire and when you next generate the changelog, it will get injected into the `index.md` file.

### What gets included?

The changelog script will include all PRs that meet the following criteria:
- they are in one of the following repositories:
  - https://github.com/gitpod-io/gitpod
  - https://github.com/gitpod-io/website
  - https://github.com/gitpod-io/workspace-images
  - let me know what more repos we should include! 
- **the PR has a release note**
- the PR is merged within the specified changelog time frame (usually first day - last day of month)
  - for https://github.com/gitpod-io/gitpod this is not the case, we consider the time that the `deployed` label was added to the merged PR

There are some more conditions for omitting PRs, which can be found in [`getPrs.js`](https://github.com/gitpod-io/website/blob/3e0ccf380d1c1d8d790589d1fe2073deaf4c030a/scripts/changelog/lib/getPrs.js#L72-L89).

### Release note blocks

Currently, the release notes blocks on PRs look a little bit like this:

````md
EPIC PR CONTENT

## Release Notes

```release-note
This PR introduces random rainbow flashes within the workspace to cater to avid gamers with the need for #RGB.
```

````

This has been the case for a long time and if it makes you happy, you're free to continue using it (although we hopefully remove it from our PR templates soon enough). There were issues with this approach, including code blocks, the inability to add multi-line content as well as generally bad readability when reviewing (the markdown inside of the block wouldn't render, so it was harder to check for link errors etc.) -> this leads us to the brand new release note blocks. They look like this:

```md

EPIC PR CONTENT - STILL HERE AND EVEN MORE **EPIC**!

## Release Notes

This PR introduces random rainbow flashes within the workspace to cater to avid gamers with the need for #RGB.

You can enable this feature by going to `Preferences` -> `Appearance` and turn on `Epic gamer mode`.

As you can see this is multi-line and 

I can add any content right until

the next level 2 heading

### This is a part of the changelog, still!

## This won't be a part of the release note

```

So to recap: you can use whichever one you prefer - this change is and most likely will be backwards-compatible. I'd love to hear feedback on this, since there may be use cases I didn't think of 👀.

### Other changes 

- this PR adds a GitHub workflow, which updates any PRs with https://github.com/gitpod-io/website/labels/changelog-update every night at UTC midnight to reflect the latest updates and remove the need to update the changelog manually. This workflow can also be run on-demand from the <kbd>Actions</kbd> tab.

- this PR adds tests to the repo using [Jest](https://jestjs.io/). It covers most of the functions that the changelog uses, and hopefully will be expanded to the whole `website` repo.
- this PR adds a help menu to the CLI with a bunch of cool flags and options. You can view the help menu by executing `node scripts/changelog -h`

## Todo

- [x] Make the PR author always the first contributor avatar (https://github.com/gitpod-io/website/pull/2760#discussion_r982221982) b46cd209ed619da5cfa9ec32f8bfe7cec6c50afc
- [x] Add multiple repo support (https://github.com/gitpod-io/website/issues/1859) [`2d7fc27`](https://github.com/gitpod-io/website/pull/2763/commits/2d7fc27b31a43c4fafaaeb85708a2611664825f6)
- [x] Only consider PRs that were `deployed` during the last month (take into account something like last 3 months)  https://github.com/gitpod-io/website/pull/2763/commits/147c9c372db88d56525208222523c6ba0700d900
- [x] Filter out self-hosted from PRs https://github.com/gitpod-io/website/pull/2763/commits/aba85f847761a2bd1d4b120b523684a1ae10748d
- [x] Make sure the [raw changelog endpoint](https://github.com/gitpod-io/website/blob/main/src/routes/changelog/raw-markdown.ts) is set up for the new release format

## Related Issue(s)

Some minor changes in this PR were first introduced in https://github.com/gitpod-io/website/pull/2759

## How to test
Try running the following:
```
CHANGELOG_GITHUB_ACCESS_TOKEN=YOUR_TOKEN node ./scripts/changelog
```

It should not require any other arguments and should produce a complete changelog file for the current month.

## Release Notes

```release-note
NONE
```

## Documentation

<a href="https://gitpod-staging.com/#https://github.com/gitpod-io/website/pull/2763"></a>

<a href="https://gitpod.io/#https://github.com/gitpod-io/website/pull/2763"></a>

### CODE REVIEW

File: scripts/generate-changelog.js [hunk 1]
```diff
@@ -23,11 +27,11 @@ const parseReleaseNote = (pr) => {
...
   const releaseNoteMatch = pr.body
-    .replace(/\r\n/g, "")
+    .replace(/\r\n/g, " ")
     .match(/```release-notes?(.+?)```/);
-  if (!releaseNoteMatch || releaseNoteMatch[1].trim().toUpperCase() === "NONE")
-    return;
-  return releaseNoteMatch[1];
+  const releaseNoteText = releaseNoteMatch && releaseNoteMatch[1].trim();
+  if (!releaseNoteMatch || releaseNoteText.toUpperCase() === "NONE") return;
+  return releaseNoteText;
```
Comment (filiptronicek) [inline scripts/generate-changelog.js:34]: This fixes a bug in which release notes like this one (note the trailing `):
```
Update stable VSCode Browser to `1.71`
```

would get displayed as just 
```
Update stable VSCode Browser to `1.71
```

because the regex matched only the first three backtics.

File: scripts/generate-changelog.js [hunk 2]
```diff
@@ -88,16 +80,16 @@ const main = async () => {
...
-  }`,
-    {
-      searchQuery,
-    }
+  }`
   );
 
   const fixesAndImprovements = search.edges
     .map((edge) => edge.node)
-    // .filter(pr => pr.number === 7609) // Helps with testing
```
Comment (filiptronicek) [inline scripts/generate-changelog.js:100]: Removed this one as this PR will not appear in our results anymore and I don't think we need it.

File: scripts/lib/utils.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    .filter((login) => !["roboquat"].includes(login))
+    .sort()
+    .join(",");
+
+export const parseReleaseNote = (pr) => {
+  const releaseNoteMatch = pr.body.match(/```release-notes?(.+?)```/s);
+  const releaseNoteText = releaseNoteMatch && releaseNoteMatch[1].trim();
+  if (!releaseNoteMatch || releaseNoteText.toUpperCase() === "NONE") return;
+  return releaseNoteText;
+};
```
Comment (filiptronicek) [inline scripts/lib/utils.js:13]: This regex has been optimized and should work in more scenarios

File: scripts/lib/getPrs.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import { getMonthBoundaries, join, dateFormat } from "./dates.js";
+import minimist from "minimist";
+
+const argv = minimist(process.argv.slice(2));
+
+const byDeployed = (pr) => {
```
Comment (filiptronicek) [inline scripts/lib/getPrs.js:6]: This function is the filtering function for the `deployed` label - it checks whether it has been added onto a PR in the specified timeframe.

File: scripts/changelog/lib/utils.js [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    .filter((login) => !["roboquat", author].includes(login))
+    .sort();
+  const allParticipants = [author, ...participants].join(",");
+  return allParticipants;
+};
+
+export const parseOldReleaseNote = (pr) => {
+  const releaseNoteMatch = pr.body.match(/```release-notes?(.+?)```/s);
+  const releaseNoteText = releaseNoteMatch && releaseNoteMatch[1].trim();
+  if (!releaseNoteMatch || releaseNoteText.toUpperCase() === "NONE") return;
```
Comment (Siddhant-K-code) [inline scripts/changelog/lib/utils.js:24]: _Code Style Suggestion_ (**_non-blocking_**):

A function should not mix return statements with and without a result.

e.g.:

```js
// Ok
function foo() {
    if (condition1) {
        return true;
    }
    return false;
}
```

```js
// Bad
function bar() {
    if (condition1) {
        return;
    }
    return false;
}
```

Comment (filiptronicek) [reply scripts/changelog/lib/utils.js:25]: Thanks for the suggestion, @Siddhant-K-code! I think what I will do here is return `null` instead of `undefined`, making it more clear.

File: scripts/changelog/lib/config.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  },
+  {
+    name: "Workspace",
+    labels: ["team: workspace"],
+    partial: "workspace",
+    prefixes: [],
+    order: 0,
+    prs: [],
+  },
+  // todo(ft): Installer (self-hosted), Workspace, supervisor
```
Comment (filiptronicek) [inline scripts/changelog/lib/config.js:67]: We can add this and others post-merge in a follow-up

File: jest.config.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export default {
+  testTimeout: 20000,
+  resetMocks: true,
+  clearMocks: true,
+};
```
Comment (Siddhant-K-code) [inline jest.config.js:5]: are we using this anywhere? can we keep it somewhere in he `scripts/changelog/` directory & not in the root?

**Review by Siddhant-K-code — COMMENTED** [review]
Thanks a lot, @filiptronicek for this proposal changes. Things looking good ✨ 



I have a few Questions about this 👇🏼. Also, a request can you please the `README` with the small instructions as `Generate Changelogs` under `CLI Commands`, I can help you with that. So that, it will be easier for folks who are using it 🙏🏼 

edit: @filiptronicek, I have added the related docs & script in [**`8645768`**](https://github.com/gitpod-io/website/pull/2763/commits/864576884f88405f574b20215aba655a97359792) & [**`2fe34d4`**](https://github.com/gitpod-io/website/pull/2763/commits/2fe34d418641e684e9bced68121b6ec2709311fb) ✅

File: scripts/changelog/lib/getPrs.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    }`;
+
+  switch (repo) {
+    case "gitpod-io/gitpod":
+      const fromAdjusted = formatDate(
+        new Date(from).getTime() - 60 * 24 * 60 * 60 * 1000,
+        dateFormat,
+        "-"
+      );
+      searchQuery = `repo:${repo} is:pr is:merged merged:${fromAdjusted}..${to} sort:updated-desc label:deployed -label:release-note-none -project:gitpod-io/22`;
```
[the remark this answers is not in the document]
Comment (filiptronicek) [reply scripts/changelog/lib/getPrs.js:79]: This is the project id of self-hosted. We currently don't want SH to be in the changelog, until we further discuss it.

File: jest.config.js (hunk 7 above)
Comment (filiptronicek) [reply jest.config.js:5]: The confit I hope should be kept the same across the project when we add any unit tests anywhere else.


Think we should keep it in the root

Comment (Siddhant-K-code) [reply jest.config.js:5]: as of now, we do not have unit tests. But, yeah make sense. It would be helpful if we add any new unit tests in the future.

File: scripts/changelog/lib/config.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
+export const lineBreak = "\r\n";
+export const changelogPath = "./src/lib/contents/changelog";
+
+export const repos = [
+  "gitpod-io/gitpod",
+  "gitpod-io/website",
```
[the remark this answers is not in the document]
Comment (filiptronicek) [reply scripts/changelog/lib/config.js:6]: The website repo sometimes has user-facing changes like adding dark mode, new docs features (by that not meant writing documentation, but things like our [IDE toggle](https://github.com/gitpod-io/website/blob/main/src/lib/components/docs/ide-toggle.svelte)) 👀

File: package.json [hunk 10]
```diff
@@ -60,11 +62,13 @@
     "hast-util-heading-rank": "^2.1.0",
     "hastscript": "^7.0.2",
     "husky": "^7.0.4",
+    "jest": "^29.1.2",
     "lint-staged": "^11.2.6",
+    "markdown-yaml-metadata-parser": "^3.0.0",
```
[the remark this answers is not in the document]
Comment (filiptronicek) [reply package.json:68]: We most likely could, I just tried it but having a hard time making it work. Found https://github.com/eugeneware/remarkable-meta for the job but it just doesn't work - when I render the markdown with the plugin, I still can't for some reason get the metadata.

Feel free to give it a go if you can, hopefully you'll be luckier than me 😄.

File: scripts/changelog/lib/config.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  {
+    name: "Website",
+    labels: [],
+    partial: "website",
+    prefixes: [],
+    order: 0,
+    prs: [],
+  },
+  {
+    name: "Workspace",
```
Comment (atduarte) [inline scripts/changelog/lib/config.js:60]: Could we have a category for workspace images?

Comment (filiptronicek) [reply scripts/changelog/lib/config.js:60]: Currently, all changellogable PRs from https://github.com/gitpod-io/workspace-images go under <kbd>Workspace</kbd>, @atduarte 👀. Are you thinking of having a separate category just for workspace image updates? Maybe we could even implement sub-categories for this.

File: scripts/changelog/lib/config.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  {
+    name: "Gitpod CLI",
+    labels: ["component: gp cli"],
+    partial: "cli",
+    prefixes: ["gp-cli"],
+    order: 0,
+    prs: [],
+  },
+  {
+    name: "Server",
```
Comment (loujaybee) [inline scripts/changelog/lib/config.js:44]: This is a very generically named component. 

From a user-facing perspective, is it obvious what this component does? (to me, no). 

**Thought:** Maybe we need "descriptions" as well as sections, that could also link to our docs.

File: scripts/changelog/lib/config.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    labels: ["component: server"],
+    partial: "server",
+    prefixes: ["server"],
+    order: 0,
+    prs: [],
+  },
+  {
+    name: "Website",
```
Comment (loujaybee) [inline scripts/changelog/lib/config.js:52]: **Thought/suggestion:** Should this be "documentation" rather than "website"? 

The website hosts our documentation, but is not a component of the product.

File: src/lib/contents/changelog/_template.md [hunk 14]
```diff
@@ -14,8 +14,6 @@ TODO: Insert the featured changelog entry here.
```
Comment (loujaybee) [inline src/lib/contents/changelog/_template.md:14]: **Suggestion:** Can we automatically create a fake `.jpg` file at the same time as we create the template? Currently if you run the automation and push, the build fails and you don't get a preview. I tend to go and copy/paste an old image as a placeholder, but this often causes confusion... 😢 

Maybe we can use: https://placebear.com/ or something 😆 🐻

File: scripts/changelog/lib/config.js (hunk 12 above)
Comment (filiptronicek) [reply scripts/changelog/lib/config.js:44]: I think it's important here to indeed think about it from the perspective of the user, although making our job here quite more difficult - there will most likely not be a lot of components with 1<>1 mappings to how users interact with Gitpod.

I agree with the descriptions! <- Although let's implement that after we think about the category names first, because that's a pre-requisite for this.

File: src/lib/contents/changelog/_template.md (hunk 14 above)
Comment (filiptronicek) [reply src/lib/contents/changelog/_template.md:14]: 💯, great idea! Added this to the to-do list for https://github.com/gitpod-io/website/pull/2926 as this is non-blocking. Already have a few ideas in mind, implementing https://placebear.com/ sounds like the thing we should enforce for the changelog coming out on April 1st.

File: scripts/changelog/lib/config.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+export const changelogPath = "./src/lib/contents/changelog";
+
+export const repos = [
+  "gitpod-io/gitpod",
+  "gitpod-io/website",
+  "gitpod-io/workspace-images",
+];
+
+export const prCategories = [
+  {
```
Comment (loujaybee) [inline scripts/changelog/lib/config.js:11]: **Suggestion:** Can we have links to our docs for each category? 🤔 

**Examples:** 
* https://www.gitpod.io/docs/references/gitpod-cli
* https://www.gitpod.io/docs/integrations/jetbrains-gateway
* https://www.gitpod.io/docs/configure/workspaces
* Would have to make a decision between: https://www.gitpod.io/docs/references/ides-and-editors/vscode-browser and https://www.gitpod.io/docs/references/ides-and-editors/vscode

File: scripts/changelog/lib/config.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
+export const lineBreak = "\r\n";
+export const changelogPath = "./src/lib/contents/changelog";
+
+export const repos = [
+  "gitpod-io/gitpod",
+  "gitpod-io/website",
```
Comment (loujaybee) [inline scripts/changelog/lib/config.js:6]: **Question:** What about openvscode server, @filiptronicek ? I'm mentioning: https://github.com/gitpod-io/openvscode-server/pull/443, in this months changelog ([link](https://github.com/gitpod-io/website/pull/2883/files#diff-636265397f13495a4334fd127a806d142e464fa75c88e8e421032ec528e0cefeR42)) but there is no associated PR.

File: scripts/changelog/lib/config.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (loujaybee) [reply scripts/changelog/lib/config.js:6]: After seeing real-time the results this generates (see below - [source](https://github.com/gitpod-io/website/pull/2883/files)), there is some value to showing this content, e.g. installation guides, etc. However, [as mentioned here](https://github.com/gitpod-io/website/pull/2763/files#r1003318142), I think we should rename this to "documentation" rather than "website". 

image

File: scripts/changelog/lib/config.js (hunk 12 above)
Comment (loujaybee) [reply scripts/changelog/lib/config.js:50]: Let me speak to PM's about this topic 🙏

File: scripts/changelog/lib/config.js (hunk 16 above)
Comment (filiptronicek) [reply scripts/changelog/lib/config.js:6]: This is a good point, openvscode-server stuff **should definitely** be here. There are two main reasons why it is not:
1. The repo is very funky with all the rebasing we do nightly. You cannot keep track of changes via commit SHAs and there is no way to tell whether a change is running in Insiders or Stable on SaaS [source: a Slack thread with Jean Pierre I cannot find].
2. There is already a mechanism for adding VS Code Browser features to the changelog – _most_ of the time, there is a corresponding PR for updating the VS Code image in https://github.com/gitpod-io/gitpod. In the case of the responsive ports PR, there is https://github.com/gitpod-io/gitpod/pull/13838 with release notes - although without the deployed and team labels - hence not present in the changelog (another issue, will fix manually this time 😆)

File: scripts/changelog/lib/config.js (hunk 11 above)
Comment (atduarte) [reply scripts/changelog/lib/config.js:76]: Noticed this was already changed as part of the october changelog 🧡

**Review by jldec — APPROVED** [review]
This all looks useful and good :)

- I like the ability to keep each month's changelog page in a separate file
- I also like the idea of allowing markdown in the changlog sections in PRs 

Long term I would like to see Gitpod changelog publishing be automated using only PRs as a source - with minimal manual touch-ups for readability. 

IMO hand-written new-feature highlights should be published as short, separate, linkable blog posts on the day that each feature ships.

**Comment by lucasvaltl** [comment]
@filiptronicek is this PR still relevant? Just asking because it has been open for a while :)

**Comment by filiptronicek** [comment]
Sorry for the lack of communication about the status here on this PR, @lucasvaltl. Currently, the PR along with the changes have been put on hold, because there's a lot of very high-prio things we need to worry about now. Thankfully enough, the changelog changes are already being used for generating changelogs (for the last two at least) and so I think this PR can just wait when it's easier for PMs to take a gander at. 

cc @loujaybee

**Comment by ArthurSens** [comment]
> Sorry for the lack of communication about the status here on this PR, @lucasvaltl. Currently, the PR along with the changes have been put on hold, because there's a lot of very high-prio things we need to worry about now. Thankfully enough, the changelog changes are already being used for generating changelogs (for the last two at least) and so I think this PR can just wait when it's easier for PMs to take a gander at.
> 
> cc @loujaybee

Would it be ok to change this PR to Draft? We've been receiving [notifications](https://gitpod.slack.com/archives/C01KLC56NP7/p1671026534414809) everyday about it 😅

**Comment by filiptronicek** [comment]
Maybe one day, but certainly not now 🫡.
